jfr/src/main/java/io/netty/handler/jfr/JfrEventHandler.java

```diff
@@ -0,0 +1,375 @@
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void close(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        CloseEvent event = new CloseEvent();
+        event.begin();
+        ctx.close(promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+    @Override
+    public void deregister(ChannelHandlerContext ctx, ChannelPromise promise) throws Exception {
+        DeregisterEvent event = new DeregisterEvent();
+        event.begin();
+        ctx.deregister(promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.commit();
+        }
+    }
+
+// NOTE: Occurs so often, currently remove for perfomance reasons
+//    @Override
+//    public void channelReadComplete(ChannelHandlerContext ctx) throws Exception {
+//        ChannelReadCompleteEvent event = new ChannelReadCompleteEvent();
+//        event.begin();
+//        ctx.fireChannelReadComplete();
+//        event.end();
+//        if (event.shouldCommit()) {
+//            event.setChannel(ctx.channel());
+//            event.commit();
+//        }
+//    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        ChannelReadEvent event = new ChannelReadEvent();
+        event.begin();
+        ctx.fireChannelRead(msg);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.msg = String.valueOf(msg);
+            event.commit();
+        }
+    }
+
+// NOTE: Occurs so often, currently remove for perfomance reasons
+//    @Override
+//    public void read(ChannelHandlerContext ctx) throws Exception {
+//        ReadEvent event = new ReadEvent();
+//        event.begin();
+//        ctx.read();
+//        event.end();
+//        if (event.shouldCommit()) {
+//            event.setChannel(ctx.channel());
+//            event.commit();
+//        }
+//    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        WriteEvent event = new WriteEvent();
+        event.begin();
+        ctx.write(msg, promise);
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.msg = String.valueOf(msg);
+            event.commit();
+        }
+    }
+
+    @Override
+    public void channelWritabilityChanged(ChannelHandlerContext ctx) throws Exception {
+        ChannelWritabilityChangedEvent event = new ChannelWritabilityChangedEvent();
+        event.begin();
+        ctx.fireChannelWritabilityChanged();
+        event.end();
+        if (event.shouldCommit()) {
+            event.setChannel(ctx.channel());
+            event.isWritable = ctx.channel().isWritable();
```

Comment:
you should store the `isWritable()` before you call `ctx.fireChannelWritabilityChanged()` above as the writability state of the channel may have already changed and so the event would not reflect the state on which it was triggered

Reply:
Should there maybe be one isWritable and one wasWritable?

Reply:
sure that can be useful as well as it will allow to see if it changed.